src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+        TIdentifierNameSyntax> : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationExpressionSyntax : TExpressionSyntax
+        where TObjectCreationExpressionSyntax : TExpressionSyntax
+        where TIdentifierNameSyntax : TExpressionSyntax
+    {
+        protected abstract SyntaxNode GenerateExpressionFromOptionalParameter(IParameterSymbol parameterSymbol);
+        protected abstract SyntaxNode? GetLocalDeclarationFromDeclarator(SyntaxNode variableDecl);
+        protected abstract SyntaxNode UpdateArgumentListSyntax(SyntaxNode argumentList, SeparatedSyntaxList<SyntaxNode> arguments);
+        protected abstract bool IsClassSpecificExpression(SyntaxNode variable);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, textSpan, cancellationToken) = context;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var expression = await document.TryGetRelevantNodeAsync<TExpressionSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+            if (expression == null || CodeRefactoringHelpers.IsNodeUnderselected(expression, textSpan))
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var expressionType = semanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            // Need to special case for expressions that are contained within a parameter
+            // because it is technically "contained" within a method, but an expression in a parameter does not make
+            // sense to introduce.
+            var parameterNode = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsParameter(node));
+            if (parameterNode is not null)
+            {
+                return;
+            }
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var containingMethod = expression.FirstAncestorOrSelf<SyntaxNode>(node => generator.GetParameterListNode(node) is not null);
+
+            if (containingMethod is null)
+            {
+                return;
+            }
+
+            var containingSymbol = semanticModel.GetDeclaredSymbol(containingMethod, cancellationToken);
+            if (containingSymbol is not IMethodSymbol methodSymbol)
+            {
+                return;
+            }
+
+            var methodKind = methodSymbol.MethodKind;
+            if (methodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction or MethodKind.Constructor))
```

Comment:
can you add a comment or something indicating if there will be any different later in terms of how you handle local-funcs and constructors?  For example, i presue we won't be adding overloads for those cases.

Reply:
correct we do not add them for constructors, but I wasn't aware the lanbguage did not support overloads for local functions. I will make sure the code action does not show up in that case

Reply:
> I will make sure the code action does not show up in that case

That, or you have to generate a unique name for the overload.